server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/RabbitMQMailQueueFactory.java

```diff
@@ -58,19 +63,30 @@ public RabbitMQMailQueue createQueue(String name) {
     @Override
     public Set<RabbitMQMailQueue> listCreatedMailQueues() {
         return mqManagementApi.listCreatedMailQueueNames()
-            .map(mailQueueFactory::create)
-            .collect(Guavate.toImmutableSet());
+            .map(this::getOrElseCreateLocally)
+            .collect(ImmutableSet.toImmutableSet());
     }
 
     private RabbitMQMailQueue attemptQueueCreation(MailQueueName mailQueueName) {
         rabbitClient.attemptQueueCreation(mailQueueName);
-        return mailQueueFactory.create(mailQueueName);
+        return createLocally(mailQueueName);
     }
 
     private Optional<RabbitMQMailQueue> getQueue(MailQueueName name) {
         return mqManagementApi.listCreatedMailQueueNames()
             .filter(name::equals)
-            .map(mailQueueFactory::create)
+            .map(this::getOrElseCreateLocally)
             .findFirst();
     }
+
+    private RabbitMQMailQueue getOrElseCreateLocally(MailQueueName name) {
+        return Optional.ofNullable(mailQueueStoreLocal.get(name))
+            .orElseGet(() -> createLocally(name));
+    }
+
+    private RabbitMQMailQueue createLocally(MailQueueName name) {
+        RabbitMQMailQueue mailQueue = mailQueueFactory.create(name);
+        mailQueueStoreLocal.put(name, mailQueue);
```

Comment:
Seems vulnerable to data races (!!!)

I bet you want to initialize the gauge a single time per node.

BUT you can:
 - Initialize the mail Queue two time
 - Return different mail queues to the caller

`return mailQueueStoreLocal.computeIfAbsent(name -> mailQueueFactory.create(name));`

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-

Reply:
BTW as you expect unicity of the RabbitMQMailQueue per James server, you need to add a concurrency test as part of rabbitMQMailQueueFactoryTest (but not in the contract, as this seems specific to RabbitMQ)